src/MCPClient/lib/clientScripts/verify_checksum.py

```diff
@@ -0,0 +1,291 @@
+                    continue
+                if line.endswith(self.OKAY_STRING):
+                    continue
+                if (
+                    line.endswith(self.FAIL_STRING)
+                    or self.ZERO_STRING in line
+                    or self.IMPROPER_STRING in line
+                ):
+                    self.job.pyprint(
+                        "{}: {}".format(self.get_ext(self.hashfile), line),
+                        file=sys.stderr,
+                    )
+                if line.endswith(self.FAILED_OPEN):
+                    self.job.pyprint(
+                        "{}: {}".format(self.get_ext(self.hashfile), line),
+                        file=sys.stderr,
+                    )
+            return err.returncode
+
+    def version(self):
+        """Return version information for the command being called."""
+        try:
+            return self._call("--version")[0]
+        except subprocess.CalledProcessError:
+            return self.COMMAND
+
+    def get_command_detail(self):
+        """Provide some way for the user to get information out of the class to
+        write METS/PREMIS provenance information.
+        """
+        if not self.command_called:
+            err_str = "Unable to retrieve information about the hashsum command called by the script"
+            raise PREMISFailure(err_str)
+        return 'program="{}"; version="{}"'.format(
+            " ".join(self.command_called), self.version()
+        )
+
+    @staticmethod
+    def _decode(out):
+        """Decode string output in Py2 or Py3 and return a list of lines to be
+        parsed elsewhere.
+        """
+        try:
+            return str(out, "utf8").split("\n")
+        except TypeError:
+            return out.decode("utf8").split("\n")
+
+    @staticmethod
+    def get_ext(path):
+        """Return the extension of the checksum file provided"""
+        ext = os.path.splitext(path)[1]
+        if not ext:
+            return path
+        return ext.replace(".", "")
+
+    @staticmethod
+    def _count_lines(path):
+        """Count the number of lines in a checksum file."""
+        count = 0
+        with open(path) as hashfile:
+            for count, _ in enumerate(hashfile):
+                pass
+        # Negate zero-based count.
+        return count + 1
+
+    @staticmethod
+    def _count_files(path):
+        """Walk the directories on a given path and count the number of files.
+        """
+        return sum([len(files) for _, _, files in os.walk(path)])
+
+
+def write_premis_event_per_file(transfer_uuid, event_detail):
+    """Write a PREMIS event for every file in the transfer on success."""
+    group_type = "transfer_id"
+    event_type = "fixity_check"
+    event_outcome = "pass"
+    # Retrieve file UUIDs from the database.
+    kwargs = {"removedtime__isnull": True, group_type: transfer_uuid}
+    with transaction.atomic():
+        file_uuids = File.objects.filter(**kwargs).values_list("uuid")
```

Comment:
Since this function is called once per checksum type (so 5x), this is going to do `SELECT * FROM File WHERE...` that many times. I think some issues that have come up recently show that we need to be thinking about the case where there are 50k files, which is where extra queries like this can get a bit slow.

I think there are at least two ways to get this to one single select:

1. Pass a queryset of file objects into write_premis_event_per_file, and construct it before calling (outside the loop). This seems like a relatively quick fix that might make it into this PR.
2. Change the job structure so that it iterates over files, and writes multiple hash events per file, rather than iterating over hashes and the files per hash. I'm guessing that is probably too broad a change for this stage in the PR.

Reply:
I went with 1! https://github.com/artefactual/archivematica/pull/1333/commits/502a3c489cd1965600191936e81e35675720993b Hopefully that looks okay. 2. would be a bigger conversation (a pretty different microservice, but could definitely be a good one to discuss in future).